client/client-multi/client-c/src/kaa/platform-impl/common/encryption_utils.c

```diff
@@ -0,0 +1,186 @@
+ *  limitations under the License.
+ */
+
+#include "platform/ext_encryption_utils.h"
+#include "platform/ext_key_utils.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include <utilities/kaa_aes_rsa.h>
+
+#include <mbedtls/pk.h>
+#include <mbedtls/entropy.h>
+#include <mbedtls/ctr_drbg.h>
+#include <mbedtls/md.h>
+
+#define ENC_SESSION_KEY_LENGTH         256
+#define KAA_SIGNATURE_LENGTH           256
+#define AES_ECB_ENCRYPTION_CHUNCK_SIZE 16
+
+extern mbedtls_pk_context pk_context_;
+
+/**
+ * Contains Endpoint keys.
+ *
+ * The main purpose of the structure is
+ * caching calculated keys.
+ */
+static struct {
+    uint8_t session_key[KAA_SESSION_KEY_LENGTH];
+    uint8_t encrypted_session_key[ENC_SESSION_KEY_LENGTH];
+    uint8_t signature[KAA_SIGNATURE_LENGTH];
+    size_t  encrypted_session_key_length;
+    size_t  session_key_length;
+} keys;
+
+kaa_error_t kaa_init_session_key(void)
+{
+    /* Initialization should be performed only once */
+    static bool initialized = false;
+    if (!initialized) {
+        /* Generate session key */
+        if (init_aes_key(keys.session_key, KAA_SESSION_KEY_LENGTH)) {
+            return KAA_ERR_BADDATA;
+        }
+
+        keys.session_key_length = KAA_SESSION_KEY_LENGTH;
+        initialized = true;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+void ext_get_endpoint_session_key(uint8_t **buffer, size_t *buffer_size)
+{
+    if (!buffer || !buffer_size) {
+        return;
+    }
+    *buffer = keys.session_key;
+    *buffer_size = keys.session_key_length;
+}
+
+/* Get encrypted session key (enctypted with remote key) */
+kaa_error_t ext_get_encrypted_session_key(uint8_t **buffer, size_t *buffer_size,
+        const uint8_t *remote_key, size_t remote_key_size)
+{
+    if (!buffer || !buffer_size || !remote_key || !remote_key_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    kaa_error_t err = rsa_encrypt(remote_key, remote_key_size, keys.session_key,
+            KAA_SESSION_KEY_LENGTH, keys.encrypted_session_key);
+    if (err) {
+        goto exit;
+    }
+
+    keys.encrypted_session_key_length = ENC_SESSION_KEY_LENGTH;
+
+    *buffer = keys.encrypted_session_key;
+    *buffer_size = keys.encrypted_session_key_length;
+
+exit:
+    return err ? KAA_ERR_BADDATA : KAA_ERR_NONE;
+}
+size_t ext_get_encrypted_data_size(size_t input_size)
+{
+    if (!input_size) {
+        return 0;
+    }
+
+    if (input_size % AES_ECB_ENCRYPTION_CHUNCK_SIZE != 0) {
+        return AES_ECB_ENCRYPTION_CHUNCK_SIZE -
+               (input_size % AES_ECB_ENCRYPTION_CHUNCK_SIZE) + input_size;
+    }
+
+    return input_size + AES_ECB_ENCRYPTION_CHUNCK_SIZE;
+}
+
+kaa_error_t ext_encrypt_data(const uint8_t *input, size_t payload_size, uint8_t *output)
+{
+    if (!input || !payload_size || !output) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    /* Adding PKCS7 padding */
+    size_t enc_data_size = ext_get_encrypted_data_size(payload_size);
+    assert(enc_data_size >= payload_size);
```

Comment:
The user may want to port Kaa to a platform that does have _assert_ . I would prefer to avoid of using of the asserts until we provide it (KAA-982) on platform independent level.